Content.Server/GameObjects/Components/Kitchen/ReagentGrinderComponent.cs

```diff
@@ -0,0 +1,367 @@
+        [ViewVariables] private bool HasBeaker => _beakerContainer.ContainedEntity != null;
+
+        [ViewVariables] private BoundUserInterface? UserInterface => Owner.GetUIOrNull(ReagentGrinderUiKey.Key);
+
+
+        private bool Powered => !Owner.TryGetComponent(out PowerReceiverComponent? receiver) || receiver.Powered;
+
+        /// <summary>
+        /// Should the BoundUI be told to update?
+        /// </summary>
+        private bool _dirty = true;
+        /// <summary>
+        /// Is the machine actively doing something and can't be used right now?
+        /// </summary>
+        private bool _busy = false;
+
+
+        //YAML serialization vars
+        private int _storageCap = 16;
+        private int _workTime = 3500; //3.5 seconds, completely arbitrary for now.
+
+        public override void ExposeData(ObjectSerializer serializer)
+        {
+            base.ExposeData(serializer);    
+            serializer.DataField(ref _storageCap, "chamber_capacity", 16);
+            serializer.DataField(ref _workTime, "time", 3500);
+        }
+
+        public override void Initialize()
+        {
+            base.Initialize();
+            //A slot for the beaker where the grounds/juices will go.
+            _beakerContainer =
+                ContainerManagerComponent.Ensure<ContainerSlot>($"{Name}-reagentContainerContainer", Owner);
+
+            //A container for the things that WILL be ground/juiced. Useful for ejecting them instead of deleting them from the hands of the user.
+            _chamber =
+                ContainerManagerComponent.Ensure<Container>($"{Name}-entityContainerContainer", Owner);
+
+            if (UserInterface != null)
+            {
+                UserInterface.OnReceiveMessage += UserInterfaceOnReceiveMessage;
+            }
+
+            _audioSystem = EntitySystem.Get<AudioSystem>();
+        }
+
+        private void UserInterfaceOnReceiveMessage(ServerBoundUserInterfaceMessage message)
+        {
+            if(!Powered || _busy)
+            {
+                return;
+            }
+
+            switch(message.Message)
+            {
+                case ReagentGrinderGrindStartMessage msg:
+                    ClickSound();
+                    DoWork(user:message.Session.AttachedEntity!,isJuiceIntent:false);
+                    break;
+
+                case ReagentGrinderJuiceStartMessage msg:
+                    ClickSound();
+                    DoWork(user: message.Session.AttachedEntity!,isJuiceIntent:true);
+                    break;
+
+                case ReagentGrinderEjectChamberAllMessage msg:
+                    if(!ChamberEmpty)
+                    {
+                        ClickSound();
+                        for (var i = _chamber.ContainedEntities.Count - 1; i >= 0; i--)
+                        {
+                            EjectSolid(_chamber.ContainedEntities.ElementAt(i).Uid);
+                        }
```

Comment:
Just ForEach it?

Reply:
bruh im not gonna lose performance gains